gh-106670: Allow Pdb to move between chained exceptions (#106676)
python/cpython: Doc/library/pdb.rst

   Print the return value for the last return of the current function.

> .. pdbcommand:: exceptions [arg]

Reviewer: To keep the consistency, we should probably write `exceptions [excnumber]` instead of `exception [arg]`. That's now `clear`/`disable` documents.

   List or jump between chained exceptions.

>    If ``Pdb.post_mortem()`` is called with a chained exception, it will now allow

Reviewer: Maybe mention `pm()` as well? It's probably a more common entry for this command.

   the user to move between the chained exceptions using ``exceptions`` command to
   list exceptions, and ``exception <number>`` to switch to that exception.

   Example::

        def out():
            try:
                middle()
            except Exception as e:
                raise ValueError("reraise middle() error") from e

        def middle():
            try:
                return inner(0)
            except Exception as e:
                raise ValueError("Middle fail")

        def inner(x):
            1 / x

         out()

>    calling ``pdb.pm()`` will allow to move between exceptions:

Author: ```suggestion
   calling ``pdb.pm()`` will allow to move between exceptions::
```

   Print the return value for the last return of the current function.

> .. pdbcommand:: exceptions [excnumber]

Reviewer 2: We need to add an entry in Doc/whatsnew/3.13.rst  since this is a new feature.

Author: Isn't that what `Misc/NEWS.d/next/Library/2023-07-12-10-59-08.gh-issue-106670.goQ2Sy.rst` is for ?

Reviewer 2: Not really. News is for the release manager. What'snew is for users.

Author: Ok, I though this was auto-merged into the whatsnew. Added.

   List or jump between chained exceptions.

   When using ``pdb.pm()``  or ``Pdb.post_mortem(...)`` with a chained exception

>    instead of a traceback, it will now allow the user to move between the

Reviewer 2: ```suggestion
   instead of a traceback, it allows the user to move between the
```

   chained exceptions using ``exceptions`` command to list exceptions, and

>    ``exception <number>`` to switch to that exception.

Author: Thoughts on adding something like this ? 
```suggestion
   ``exception <number>`` to switch to that exception. In the case of ``ExceptionGroup``s this will also allow to inspect sub exceptions.
```

Reviewer 2: I don't think this feature, as is, should allow traversing into exception groups. The arg implies a linear order of exceptions, and if you have an exception group in an exception chain, and each exception in the group can also be in a chain, it's not obvious how to flatten all this into a linear order.

Author: Ok, removed.

   Example::

        def out():
            try:
                middle()
            except Exception as e:
                raise ValueError("reraise middle() error") from e

        def middle():
            try:
                return inner(0)
            except Exception as e:
                raise ValueError("Middle fail")

        def inner(x):
            1 / x

         out()

   calling ``pdb.pm()`` will allow to move between exceptions::

    > example.py(5)out()
    -> raise ValueError("reraise middle() error") from e

    (Pdb) exceptions
      0 ZeroDivisionError('division by zero')
      1 ValueError('Middle fail')
    > 2 ValueError('reraise middle() error')

    (Pdb) exceptions 0
    > example.py(16)inner()
    -> 1 / x

    (Pdb) up
    > example.py(10)middle()
    -> return inner(0)

> 

Reviewer 2: We need a versionadded entry here to indicate that this is new in 3.13.

Author: Added at the top just after `.. pdbcommand::` as all of this is new.

Reviewer 2: It typically goes in the end. You write the docs to describe the current situation and then add versionadded and versionchanged directives at the end. Look at other docs for examples.

Author: ok. Should be done.
